statie/source/_posts/2019/2019-05-21-review-of-the-refactoring.md

```diff
@@ -0,0 +1,80 @@
+
+Honestly, the important part of the book is the first 100 pages (of 400) and I recommend for every developer to read the first part of the book, the rest of the book is optional.
+
+## Catalogue of Refactorings
+
+The rest 300 pages is a catalog of refactoring and you can find it [online](https://refactoring.com/catalog/).
+Each refactoring includes motivation - when, mechanics - how - in step-by-step form, and examples.
+I like the step-by-step instructions when tests pass after each instruction.
+It is awkward to apply these steps for simple refactorings, and powerful to apply for complicated refactorings or refactorings that involves a lot of occurrences.
+
+All refactorings are explained really simply, even the difficult parts.
+That's not because oversimplification, but it's because the author is a professional teacher, the explanation is just great.
+Most of the refactorings are simple (sometimes too much like [Remove Dead Code](https://refactoring.com/catalog/removeDeadCode.html)) but still practical.
+
+## Personal Impression
+
+Almost all refactorings include inverse refactoring.
+When there is [Extract Variable](https://refactoring.com/catalog/extractVariable.html), you can find also [Inline Variable](https://refactoring.com/catalog/inlineVariable.html).
+That was a surprise for me as I thought the first option is always better (and I learned that's not true).
+
+I want to point out [Replace Type Code with Subclasses](https://refactoring.com/catalog/replaceTypeCodeWithSubclasses.html) and [Introduce Special Case](https://refactoring.com/catalog/introduceSpecialCase.html).
+Both introduce reasonable inheritance.
+Reasonable inheritance is the most difficult part of OOP for me and these refactorings taught me something new.
+
+On the other hand, most of the refactorings were pretty simple or already known to me and I'd like to see more advanced refactorings, that's what I'm missing in the book.
+
+## Language & Code
+
+The book is easy to read, the author is using easy English and anyone who can read this review can read the book.
+
+Code snippets are written in JavaScript, are simple and usually pretty short.
+I saw a couple of pages of the 1st edition of the book and I can only congratulate Fowler for choosing the JavaScript because original Java code is really verbose.
+Just compare
+
+#### 2nd edition in JavaScript
+
+```js
+movie = {'name': "Pulp fiction"};
+```
+
+#### 1st edition in Java
+
+```java
+class Movie {
+    private String name;
+    public Movie(String name) {
+        this.name = name;
+    }
+    public String getName() {
+        return name;
+    }
+}
+
+Movie movie = new Movie("Pulp Fiction");
+```
+
+this is not a Java criticism, JavaScript is just better for educative purposes in this case.
```

Comment:
Depends what you teach and who you teach ;) I personally prefer Java here, because it's closer to my natural language - PHP.